replica/table.cc

```diff
@@ -644,26 +669,38 @@ table::add_sstables_and_update_cache(const std::vector<sstables::shared_sstable>
 future<>
 table::update_cache(compaction_group& cg, lw_shared_ptr<memtable> m, std::vector<sstables::shared_sstable> ssts) {
     auto permit = co_await seastar::get_units(_sstable_set_mutation_sem, 1);
-    mutation_source_opt ms_opt;
-    if (ssts.size() == 1) {
-        ms_opt = ssts.front()->as_mutation_source();
-    } else {
-        std::vector<mutation_source> sources;
-        sources.reserve(ssts.size());
-        for (auto& sst : ssts) {
-            sources.push_back(sst->as_mutation_source());
+
+    class updater : public table_sstables_adder {
+        lw_shared_ptr<memtable> _m;
+        mutation_source_opt _ms_opt;
+    public:
+        updater(table& t, compaction_group& cg, lw_shared_ptr<memtable> m, std::vector<sstables::shared_sstable> ssts)
+            : table_sstables_adder(t, cg, std::move(ssts), is_main::yes)
+            , _m(std::move(m))
+        {
+            if (_sstables.size() == 1) {
+                _ms_opt = _sstables.front()->as_mutation_source();
```

Comment:
Let's move this optimization to make_combined_mutation_source() (can be done later)

Reply:
makes sense